starlette/routing.py

```diff
@@ -572,6 +574,91 @@ def __call__(self: _T, app: object) -> _T:
         return self
 
 
+class LifespanException(Exception):
+    def __init__(self, message: str):
+        self.message = message
+
+
+@asynccontextmanager
+async def _app_lifespan(scope: Scope, app: ASGIApp) -> typing.AsyncIterator[None]:
+    startup_sent = Event()
+    startup_done = Event()
+
+    shutdown_init = Event()
+    shutdown_sent = Event()
+    shutdown_done = Event()
+
+    lifespan_supported = False
+    exception = None
+
+    async def receive() -> Message:
+        nonlocal lifespan_supported
+
+        lifespan_supported = True
+
+        if not startup_sent.is_set():
+            startup_sent.set()
+            return {"type": "lifespan.startup"}
+
+        elif startup_done.is_set() and not shutdown_sent.is_set():
+            await shutdown_init.wait()
+            shutdown_sent.set()
+            return {"type": "lifespan.shutdown"}
+
+        else:
+            raise RuntimeError("unexpected receive")
+
+    async def send(message: Message) -> None:
+        nonlocal exception, lifespan_supported
+
+        lifespan_supported = True
+
+        if startup_sent.is_set() and not startup_done.is_set():
+            if message["type"] == "lifespan.startup.complete":
+                pass
+            elif message["type"] == "lifespan.startup.failed":
+                exception = message.get("message", "")
+            else:
+                raise ValueError(f"unexpected type: {message['type']}")
+            startup_done.set()
+
+        elif shutdown_sent.is_set() and not shutdown_done.is_set():
+            if message["type"] == "lifespan.shutdown.complete":
+                pass
+            elif message["type"] == "lifespan.shutdown.failed":
+                exception = message.get("message", "")
+            else:
+                raise ValueError(f"unexpected type: {message['type']}")
+            shutdown_done.set()
+
+        else:
+            raise RuntimeError("unexpected send")
+
+    # This wrapper is needed because TaskGroup.start_soon does not like that
+    # App returns Awaitable instead of Coroutine
```

Comment:
Yup that is the case. If we use this somewhere else in the library (I don't recall) it might be worth doing something more generic since the issue relates to awaitables/coroutines and not specifically ASGI apps.

Reply:
Yeah makes sense, for now this wrapper turned out to be a good thing because I added some code at the end to check if the app did not return in a situation where that would block the context manager.